internal/server/event_handlers.go

```diff
@@ -251,19 +253,55 @@ func (s *EventServer) getNotificationParams(ctx context.Context, event *eventv1.
 		return nil, nil, "", 0, nil
 	}
 
-	sender, token, err := createNotifier(ctx, s.kubeClient, provider)
+	// Create a copy of the event and combine event metadata
+	notification := *event.DeepCopy()
+	s.combineEventMetadata(ctx, &notification, alert)
+
+	// Create a commit status for the given provider and event, if applicable.
+	commitStatus, err := createCommitStatus(ctx, provider, &notification, alert)
 	if err != nil {
-		return nil, nil, "", 0, fmt.Errorf("failed to initialize notifier for provider '%s': %w", provider.Name, err)
+		logger.Error(err, "failed to evaluate commitStatusExpr", "expression", provider.Spec.CommitStatusExpr, "provider", provider.Name)
+		s.Eventf(&provider, corev1.EventTypeWarning, "CommitStatusExprEvalFailed", "failed to evaluate commitStatusExpr %q for provider %s: %s", provider.Spec.CommitStatusExpr, provider.Name, err)
+		return nil, nil, "", 0, fmt.Errorf("failed to create commit status: %w", err)
 	}
 
-	notification := *event.DeepCopy()
-	s.combineEventMetadata(ctx, &notification, alert)
+	sender, token, err := createNotifier(ctx, s.kubeClient, provider, commitStatus)
+	if err != nil {
+		return nil, nil, "", 0, fmt.Errorf("failed to initialize notifier for provider '%s': %w", provider.Name, err)
+	}
 
 	return sender, &notification, token, provider.GetTimeout(), nil
 }
 
+// createCommitStatus creates a commit status for the given provider and event.
+// If the provider has a commitStatusExpr, it will be used to compute a commit status.
+// Otherwise, a default commit status will be generated using the Provider UID and event metadata.
+// If the provider is not a git provider, the commit status will be an empty string.
+// If the commitStatusExpr fails to compile or is invalid, an error will be returned.
+func createCommitStatus(ctx context.Context, provider apiv1beta3.Provider, event *eventv1.Event, alert *apiv1beta3.Alert) (commitStatus string, err error) {
+	if !isGitProvider(provider.Spec.Type) {
+		return "", nil
+	}
+
+	if provider.Spec.CommitStatusExpr != "" {
+		err = validateCommitStatusExpr(provider.Spec.CommitStatusExpr)
```

Comment:
It seems like the existence of this function doesn't add much value to the code to me, we could just call `newCommitStatusExpression()` here and pass the built `*cel.Expression` to `newCommitStatus()`, which avoids building the expression twice. Then the tests currently written for `validateCommitStatusExpr()` would become tests for `newCommitStatusExpression()`